src/svd/device.py

```diff
@@ -0,0 +1,1585 @@
+        address_offset = element.offset
+
+        registers: Optional[Mapping[str, _RegisterSpec]] = None
+        fields: Optional[Mapping[str, _FieldSpec]] = None
+
+        if isinstance(element, bindings.RegisterElement):
+            # Register addresses are defined relative to the enclosing element
+            if address_offset is not None:
+                address_start = base_address + address_offset
+            else:
+                address_start = base_address
+
+            size_bytes = reg_props.size // 8
+
+            # Contiguous fill
+            if dim_props is None or dim_props.step == size_bytes:
+                length = dim_props.length if dim_props is not None else 1
+                address_end = address_start + size_bytes * length
+                memory.fill(
+                    start=address_start,
+                    end=address_end,
+                    content=reg_props.reset_value,
+                    item_size=size_bytes,
+                )
+
+            # Fill with gaps
+            elif dim_props is not None and dim_props.step > size_bytes:
+                memory.fill(
+                    start=address_start,
+                    end=address_start + size_bytes,
+                    content=reg_props.reset_value,
+                    item_size=size_bytes,
+                )
+                memory.tile(
+                    start=address_start,
+                    end=address_start + dim_props.step,
+                    times=dim_props.length,
+                )
+
+            else:
+                raise SvdDefinitionError(
+                    element,
+                    f"step of 0x{dim_props.step:x} is less than the size of the "
+                    f"array (0x{size_bytes})",
+                )
+
+            fields = _extract_field_descriptions(element.fields)
+
+        else:  # ClusterElement
+            # By the SVD specification, cluster addresses are defined relative to the peripheral
+            # base address, but some SVDs don't follow this rule.
+            if address_offset is not None:
+                address_start = base_address + address_offset
+                # address = address_offset
+            else:
+                address_start = base_address
+
+            (
+                sub_descriptions,
+                sub_min_address,
+                sub_max_address,
+            ) = _extract_register_descriptions_helper(
+                memory=memory,
+                elements=element.registers,
+                base_reg_props=reg_props,
+                base_address=address_start,
+            )
+
+            if sub_descriptions:
+                registers = {d.name: d for d in sub_descriptions}
+
+                if address_offset is None:
+                    address_start = sub_min_address
+
+                if dim_props is not None and dim_props.length > 1:
+                    # FIXME: should this use sub_min_address? address of the cluster doesn't *really* matter
```

Comment:
I don't know if this offers much, but IMO `sub_max_address` makes sense here since we're checking element coverage over its range. Because of the assignment a couple lines up, sing `sub_min_address` risks this always returning False even if the step is positive and the max != address start.